.crawler/src/known_network.rs

```diff
@@ -26,37 +28,90 @@ use crate::{
+#[derive(Clone)]
+#[allow(dead_code)]
+pub struct NetworkSummary {
+    // The number of all known nodes.
+    num_known_nodes: usize,
+    // The number of all known connections.
+    num_known_connections: usize,
+    // The number of nodes that haven't provided their state yet.
+    nodes_pending_state: usize,
+    // The types of nodes and their respective counts.
+    types: HashMap<NodeType, usize>,
+    // The versions of nodes and their respective counts.
+    versions: HashMap<u32, usize>,
+    // The node states of nodes and their respective counts.
+    states: HashMap<State, usize>,
+    // The heights of nodes and their respective counts.
+    heights: HashMap<u32, usize>,
+}
+
+impl fmt::Debug for NetworkSummary {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Network summary")
+            .field("number of known nodes", &self.num_known_nodes)
+            .field("number of known connections", &self.num_known_connections)
+            .field("nodes pending state", &self.nodes_pending_state)
+            .field("types", &self.types)
+            .field("versions", &self.versions)
+            .field("states", &self.states)
+            .finish()
+    }
+}
+
 /// Node information collected while crawling.
 #[derive(Debug, Clone)]
 pub struct NodeMeta {
     #[allow(dead_code)]
     listening_addr: SocketAddr,
-    last_height: Option<u32>,
-    // Set on disconnect.
-    last_crawled: Option<OffsetDateTime>,
-
-    // Useful for keeping track of crawl round state.
+    // The details of the node's state.
+    pub state: Option<NodeState>,
+    // The last connection success/failure timestamp.
+    timestamp: Option<OffsetDateTime>,
+    // The number of lists of peers received from the node.
     received_peer_sets: u8,
+    // The number of subsequent connection failures.
+    connection_failures: u8,
 }
 
 impl NodeMeta {
     fn new(listening_addr: SocketAddr) -> Self {
         Self {
             listening_addr,
-            last_height: None,
-            last_crawled: None,
+            state: None,
+            timestamp: None,
             received_peer_sets: 0,
+            connection_failures: 0,
         }
     }
 
     fn reset_crawl_state(&mut self) {
         self.received_peer_sets = 0;
-        self.last_crawled = Some(OffsetDateTime::now_utc());
+        self.connection_failures = 0;
+        self.timestamp = Some(OffsetDateTime::now_utc());
     }
 
     fn needs_refreshing(&self) -> bool {
-        if let Some(last_crawled) = self.last_crawled {
-            (OffsetDateTime::now_utc() - last_crawled).whole_minutes() > CRAWL_INTERVAL_MINS
+        if let Some(timestamp) = self.timestamp {
+            let crawl_interval = if self.state.is_some() {
+                CRAWL_INTERVAL_MINS
+            } else {
+                // Delay further connection attempts to nodes that are hard to connect to.
+                self.connection_failures as i64
```

Comment:
Why is the number of connection failures being used as the crawl interval in this case? Is it just to have some non-zero value to use as an interval if a peer is hard to connect to?

Reply:
if we can't connect to some node, it's just a simple way to retry more quickly than in the success case (as we want to quickly reconnect to the temporarily unavailable nodes), but with longer and longer delays in case it keeps failing